### PR TITLE
Common buffer for load balancing

### PR DESCRIPTION
**Pull request by ChasingNeutrons**
With help from @valeriaRaffuzzi , I made a common buffer (i.e., non-thread private) for fixed source problems where there's significant multiplicity. For example, near-critical problems and problems with variance reduction + weight windows.
There's also a few typo fixes about!

### CODE REVIEW

**Review by valeriaRaffuzzi — APPROVED** [review]
This works for me! 
It's totally needed for calculations with weight windows, makes things much faster and parallel performance much better.

File: PhysicsPackages/fixedSourcePhysicsPackage_class.f90 [hunk 1]
```diff
@@ -189,11 +192,31 @@ subroutine cycles(self, tally, N_cycles)
...
+            cycle
+          end if
+
+          ! Clear out common queue
+          if (associated(self % commonBuffer)) then
+            !$omp critical
+            if (.not. self % commonBuffer % isEmpty()) then
+              call self % commonBuffer % releaseCritical(p)
+            end if
+            !$omp end critical
```
Comment (Mikolaj-A-Kowalski) [inline PhysicsPackages/fixedSourcePhysicsPackage_class.f90:216]:
```suggestion
            if (.not. self % commonBuffer % isEmpty()) then
              call self % commonBuffer % releaseCritical(p)
            end if
```
The critical region is controlled inside the subroutine isn't it?

File: PhysicsPackages/fixedSourcePhysicsPackage_class.f90 [hunk 2]
```diff
@@ -189,11 +192,31 @@ subroutine cycles(self, tally, N_cycles)
             if(p % isDead) exit history
           end do history
 
+          ! If buffer is quite full, shift some particles to the commonBuffer
```
Comment (Mikolaj-A-Kowalski) [inline PhysicsPackages/fixedSourcePhysicsPackage_class.f90:195]: 'quite full'  :gb: :tea: :crown:

**Review by Mikolaj-A-Kowalski — APPROVED** [review]
Thanks for the change! Just the one minor suggestion and I think it is ready to merge.

File: PhysicsPackages/fixedSourcePhysicsPackage_class.f90 (hunk 1 above)
Comment (ChasingNeutrons) [reply PhysicsPackages/fixedSourcePhysicsPackage_class.f90:216]: The thing I'm being careful about here is that check on whether the buffer is empty. If there are two threads and one particle in the buffer, both will enter the conditional, the first will critically remove the particle, while the second will then be left trying to pop a particle from an empty buffer.
We could put a conditional inside the release as a check in principle? But we would still have the critical in the physics package too. What do you think?

Comment (Mikolaj-A-Kowalski) [reply PhysicsPackages/fixedSourcePhysicsPackage_class.f90:216]: Good point! It completely eluded me. I guess the only thing that is worth changing here is maybe adding a comment to explain why it is like that. I am sure I am not a last person who might be confused by it :-)!

Comment (ChasingNeutrons) [reply PhysicsPackages/fixedSourcePhysicsPackage_class.f90:216]: Hear hear... We only found out because we were bitten by it ourselves!

**Pull request merged**